agent/discover/cluster.go

```diff
@@ -0,0 +1,91 @@
+package discover
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/ClusterLabs/ha_cluster_exporter/collector/pacemaker/cib"
+	"github.com/ClusterLabs/ha_cluster_exporter/collector/pacemaker/crmmon"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+	consul_internal "github.com/trento-project/trento/internal/consul"
+)
+
+const crm_monPath string = "/usr/sbin/crm_mon"
+const cib_admPath string = "/usr/bin/cibadmin"
+
+//const crm_monPath string = "./test/fake/crm_mon.sh"
+//const cib_admPath string = "./test/fake_cibadmin.sh"
+
+type ClusterDiscoverInterface func(client consul.Client) ClusterDiscover
+
+type ClusterDiscover struct {
+	crmMonParser   crmmon.Parser
+	cibConfig      cib.Root
+	clusterName    string
+	stonithEnabled bool
+	stonithType    consul_internal.ClusterStonithType
+}
+
+func (cluster ClusterDiscover) shouldDiscover(client consul.Client) bool {
+	// ### Check if we have cibadmin available
+	return true
+}
+
+func (cluster ClusterDiscover) KVStorePrefix() string {
+	return fmt.Sprintf(
+		"%s/%s",
+		consul_internal.KV_CLUSTERS_PATH, cluster.clusterName)
+}
+
+func NewClusterDiscover(client consul.Client) ClusterDiscoverInterface {
```

Comment:
A dumb question. Why do we need here this `Interface` function that returns a function?
It looks pretty strange

Reply:
I don't really know. I just copied this pattern from the Checker. there was no architecture discussion on that, it was just merged, so I assumed it is good practice.